### PR TITLE
Change ordering of array for y-axis data

### PR DESCRIPTION
**Pull request by AUTProgram**
This PR makes a major breaking change to ionics_fits as it changes the ordering of axes in arrays for y-axis data. The new ordering is `(num_y_channels, num_samples)`. This fixes the problems mentioned in https://github.com/OxIonics/ionics_fits/issues/97, see also that issue for a more general discussion.

The main code changes resulting from this decision are as follows:
-) Model functions with more than one y-channel now return y-data array with new shape
-) Fitter now requires observed y-data to be passed correctly with new shape and assumes that all models return data in the new shape
-) Function annotations for all models and fitter functions have been updated
-) Because only one independent variable is currently supported, the fitter always stores x-axis data as a 1D array. Even in the previous version, an error was thrown whenever the fitter was passed x-axis data that was not a 1D array. Since one or more y-channels (and corresponding sigma-channels) are supported, the fitter internally always uses and stores y-axis data (and sigma) data as 2D arrays.
-) As a result, y-data passed from fitter to models for parameter estimation are 2D, even for models that only contain a single y-channel. Therefore, in `estimate_parameters` for all models with just a single y-channel, have added a line to squeeze the y-array to make sure it is 1D.
-) In order to be able to use results from fitting for plotting, the `Fitter.evaluate()` method now takes an optional keyword argument called `transpose_and_squeeze`. This ensures that arrays returned are compatible with the `plot()` function of matplotlib. Squeezing is required because `plot()` requires data to be contained along columns and does not produce the desired result when using arrays of shape `(1, n_samples)`.

Apart from these changes, I have also modified the `Model.get_num_y_channels()` method to raise `NotImplementedError` to ensure that subclasses define this method. I ran into a problem where `MappedModel` for a model with multiple y-channels did not return the correct number of channels because the mapped model did not call `inner.get_num_y_channels()` and instead silently took the default of 1 from `Model.get_num_y_channels()`.

I have run `poe test`, all tests pass successfully.

One remaining issue is that strictly speaking, our function annotation for `Fitter.__init__()` is not quite correct. We currently allow the y-axis data for these models to be either 1D or 2D, but the annotation `y: ArrayLike[("num_y_channels", "num_samples"), np.float64]`, says it should only be 2D. Similarly, the annotations for `estimate_parameters` methods incorrectly say they only expect 1D arrays.

### CODE REVIEW

**Comment by AUTProgram** [comment]
On a different note, it's not clear to me whether the option of passing an `int` into `Fitter.evaluate()` as `x_fit` is actually helpful. I assume people will only rarely use this functionality and in those cases, they could also just do `x_fit = np.linspace(np.min(fitter.x), np.max(fitter.x), n_steps)` themselves. Conceptually, the purpose of `x_fit` is currently very different depending on whether the user passes an array or an integer and it seems to be causing problems for type checking too.

File: ionics_fits/common.py [hunk 1]
```diff
@@ -737,17 +732,19 @@ def _fit_significance(self) -> Optional[float]:
...
+        Array[("num_y_channels", "num_samples"), np.float64],
     ]:
         """Evaluates the model function using the fitted parameter set.
 
         :param x_fit: optional x-axis points to evaluate the model at. If `None` we use
             the dataset values. If a scalar we generate an axis of linearly spaced
-            points between the minimum and maxim value of the x-axis dataset. Otherwise
-            it should be an array of x-axis data points to use.
+            points between the minimum and maximum value of the x-axis dataset.
+            Otherwise it should be an array of x-axis data points to use.
```
Comment (hartytp) [inline ionics_fits/common.py:747]: Can you document `transpose_and_squeeze` please? We should explicitly mention plotting here. e.g. say one can do `plt.plot(*fit.eval(True))`

Should we put this before `x_fit`? It is a breaking change which is bad, bit it makes the most common use-case (plotting) easier since the user does not need to do `evaluate(None, True)`.

Could you update the start of the README which has a plotting example please? And also check for uses within `ionics_fits` of `evaluate` which may need an update (I think it's used in the tests).

File: ionics_fits/common.py [hunk 2]
```diff
@@ -757,8 +754,12 @@ def evaluate(
             x_fit = np.linspace(np.min(self.x), np.max(self.x), x_fit)
 
         y_fit = self.model.func(x_fit, self.values)
-        return x_fit, y_fit  # type: ignore
 
-    def residuals(self) -> Array[("num_samples", "num_y_channels"), np.float64]:
+        if transpose_and_squeeze:
+            return x_fit, y_fit.T.squeeze()  # type: ignore
+        else:
```
Comment (hartytp) [inline ionics_fits/common.py:760]: I'd drop the `else` here?

File: ionics_fits/models/containers.py [hunk 3]
```diff
@@ -69,28 +69,31 @@ def func(
...
 
     def calculate_derived_params(
         self,
         x: Array[("num_samples",), np.float64],
-        y: Array[("num_samples",), np.float64],
+        y: Array[("num_y_channels", "num_samples"), np.float64],
         fitted_params: Dict[str, float],
         fit_uncertainties: Dict[str, float],
     ) -> Tuple[Dict[str, float], Dict[str, float]]:
+        y = np.atleast_2d(y)
```
Comment (hartytp) [inline ionics_fits/models/containers.py:95]: why is this needed?

Comment (hartytp) [reply ionics_fits/models/containers.py:95]: (isn't it already done by the fitter? or is the idea that the model may be used without the fitter?)

File: ionics_fits/models/gaussian.py [hunk 4]
```diff
@@ -70,6 +74,9 @@ def estimate_parameters(
         :param model_parameters: dictionary mapping model parameter names to their
             metadata, rescaled if allowed.
         """
+        # Ensure that y is a 1D array
+        y = np.squeeze(y)
```
Comment (hartytp) [inline ionics_fits/models/gaussian.py:78]: why is this needed?

**Comment by hartytp** [comment]
> -) Model functions with more than one y-channel now return y-data array with new shape

Is this true? Doesn't the sinusoid model return a 1D array?

**Comment by hartytp** [comment]
> On a different note, it's not clear to me whether the option of passing an int into Fitter.evaluate() as x_fit is actually helpful. I assume people will only rarely use this functionality and in those cases, they could also just do x_fit = np.linspace(np.min(fitter.x), np.max(fitter.x), n_steps) themselves and pass it to the function. Conceptually, the purpose of x_fit is currently very different depending on whether the user passes an array or an integer and it seems to be causing problems for type checking too.

Yes, I agree with this. If you want to remove that option go ahead!

**Comment by AUTProgram** [comment]
> Is this true? Doesn't the sinusoid model return a 1D array? Or by "new shape" do you mean swapping num_samples and num_y_channels?

Yeah, I only meant swapping the order to `(num_y_channels, num_samples)`. But it's only relevant for models with **more** than one y-channel. Models with a single y-channel still return a 1D array if the independent variable `x` is 1D (but note that the user can in principle also pass 2D arrays as `x` to models with a single y-channel, as long as one of the axes has length one).

**Comment by AUTProgram** [comment]
> Is this needed? What would have broken if you didn't do it? Doesn't numpy broadcasting mean this is fine?

Yes, I didn't add this at first, but then almost all tests fail at the `estimate_parameters` stage. The problem isn't so much the broadcasting, the problem is that some of the algorithms used to find initial estimates implicitly assume that the x- and y-data is 1D. For example, in `get_spectrum`, we do `y_f = np.fft.fft(y, norm="ortho") / np.sqrt(n)` and then afterwards `y_f = y_f[1:]`. But since `y_f` now has shape `(1, num_samples)`, the second expression results in `y_f = []`. If `y_f` was always 2D, we could fix this by doing `y_f = y_f[:, 1:]`. But this will fail if the user passes a 1D array since then the second index would be out of bounds, because the second axis doesn't exist. As another example, in `gaussian.estimate_parameters`, we do `model_parameters["y0"].heuristic = y[0]`. But if `y` is 2D as passed by the fitter, the heuristic for `y0` is no longer a scalar, but just the whole y-array! Similar problems occcur for the other models.

In short, the problem is that even models that contain only a single y-channel, in particular their `estimate_parameters` methods, need to handle both 1D and 2D y-arrays, because the fitter internally always uses 2D arrays and calls methods of `Model`. There's two ways out of this: 1) promote everything to 2D arrays in models too, as we do in fitter 2) internally squeeze any 2D arrays to 1D in `estimate_parameters` to get valid estimates. I prefer option 2) because models with a single y-channel conceptually just have one dependent variable and I find it more intuitive to think about this is as a 1D array. Besides, if we ever started supporting two independent variables, we would we then have to internally promote everything to 3D arrays in models, which just seems unnecessary.

File: ionics_fits/models/gaussian.py (hunk 4 above)
Comment (AUTProgram) [reply ionics_fits/models/gaussian.py:78]: See comments in main thread.

File: ionics_fits/common.py (hunk 1 above)
Comment (AUTProgram) [reply ionics_fits/common.py:747]: I'd be in favour of keeping the signature as it is currently, people can already do `plt.plot(*fit.eval(tranpose_and_squeeze=True))`. I would say the main parameter of this method is `x_fit` and would put this as the first parameter therefore.

File: ionics_fits/models/containers.py (hunk 3 above)
Comment (AUTProgram) [reply ionics_fits/models/containers.py:95]: I don't think it's actually needed, will take out

**Comment by AUTProgram** [comment]
@hartytp Feel free to have another look after latest updates, otherwise I'll merge today after lunch

File: ionics_fits/common.py (hunk 1 above)
Comment (hartytp) [reply ionics_fits/common.py]: > I would say the main parameter of this method is x_fit and would put this as the first parameter therefore.

Well, I guess this is the main question. To me the most common use case is `plt.plot(*fit.eval(transpose_and_squeeze=True))` rather than providing `x_fit`, but if that's not the case I agree with you. And, even when one does provide `x_fit` one is probably still setting `transpose_and_squeeze` to `True`.

What I don't like about the current structure is that the most common use of this method requires me to remember a somewhat verbose argument name, which is a shame.

**Comment by hartytp** [comment]
@AUTProgram modulo the small question about the arguments in `evaluate` I'm very happy for this to be merged.

File: ionics_fits/common.py (hunk 1 above)
Comment (hartytp) [reply ionics_fits/common.py]: > Since the default for transpose_and_squeeze is True

I hadn't appreciated this. I think this is a bad idea.

IMHO it's really important for libraries to behave consistently internally, this means that the default shapes used by functions should all be the same. Having options to reshape is useful, but I think it should never be default as it's just too much of a footgun.
